howtos/diffs/lr-schedule.diff

```diff
@@ -0,0 +1,90 @@
+diff --git a/examples/mnist/train.py b/examples/mnist/train.py
+index 5dc74db..bf4158a 100644
+--- a/examples/mnist/train.py
++++ b/examples/mnist/train.py
+@@ -84,6 +84,16 @@ def create_model(key):
+   model = nn.Model(CNN, initial_params)
+   return model
+ 
++def create_triangular_schedule(lr_min, lr_max, epochs):
++  """Return fn from epoch to LR, linearly interpolating between `lr_min`->`lr_max`->`lr_min`."""
++  top = epochs // 2
++  def learning_rate_fn(epoch):
++    if epoch < top:
++      lr = lr_min + epoch/top * (lr_max - lr_min)
++    else:
++      lr = lr_max - ((epoch - top)/top) * (lr_max - lr_min)
++    return lr
++  return learning_rate_fn
+ 
+ def create_optimizer(model, learning_rate, beta):
+   optimizer_def = optim.Momentum(learning_rate=learning_rate, beta=beta)
+@@ -111,7 +121,7 @@ def compute_metrics(logits, labels):
+ 
+ 
+ @jax.jit
+-def train_step(optimizer, batch):
++def train_step(optimizer, batch, lr):
+   """Train for a single step."""
+   def loss_fn(model):
+     logits = model(batch['image'])
+@@ -119,7 +129,7 @@ def train_step(optimizer, batch):
+     return loss, logits
+   grad_fn = jax.value_and_grad(loss_fn, has_aux=True)
+   (_, logits), grad = grad_fn(optimizer.target)
+-  optimizer = optimizer.apply_gradient(grad)
++  optimizer = optimizer.apply_gradient(grad, learning_rate=lr)
+   metrics = compute_metrics(logits, batch['label'])
+   return optimizer, metrics
+ 
+@@ -130,18 +140,19 @@ def eval_step(model, batch):
+   return compute_metrics(logits, batch['label'])
+ 
+ 
+-def train_epoch(optimizer, train_ds, batch_size, epoch, rng):
++def train_epoch(optimizer, train_ds, batch_size, epoch, rng, learning_rate_fn):
```

Comment:
Actually sorry for just now pointing this out, but let's pass `lr` (`learning_rate_fn(epoch)`) as an argument here instead of the learning rate function.